openTSNE/affinity.py

```diff
@@ -255,9 +257,26 @@ def check_perplexity(self, perplexity):
 def build_knn_index(
     data, method, k, metric, metric_params=None, n_jobs=1, random_state=None
 ):
+    if not sp.issparse(data) and metric in [
+        "cosine",
+        "euclidean",
+        "manhattan",
+        "hamming",
+        "dot",
+        "l1",
+        "l2",
+        "taxicab",
+    ]:
+        approx_method = nearest_neighbors.Annoy
+    else:
+        approx_method = nearest_neighbors.NNDescent
+
     methods = {
         "exact": nearest_neighbors.BallTree,
-        "approx": nearest_neighbors.NNDescent,
+        "auto": approx_method,
```

Comment:
I know we've already discussed this before elsewhere, and you had pointed out that Annoy is just as fast as exact search for smaller data sets. Even so, wouldn't it be better to prefer exact search wherever possible just from the correctness point of view? Then "approx" might switch between annoy and nndescent, and "auto" between all three. This doesn't need to be fixed here and we can put this in once a cutoff is determined.

Reply:
Sure. How about I put a 1000 threshold for now, and you can adapt it later if you want? By the way, the same reasoning would suggest to implement exact t-SNE and actually default to that for very small datasets. I think the lack of exact t-SNE implementation (that I guess could be nicely parallelized too) is a shortcoming of openTSNE.

Reply:
That works. Yes, I have been thinking of adding exact t-SNE, but I'm still figuring out how to incorporate it into the current framework. This won't go in this release though.

Reply:
OK done. Great to hear that it's on the to-do!